Mock device

msudumbrekar-TT opened the pull request:
# Mock Device: 
Issue : https://github.com/tenstorrent/tt-metal/issues/14000
UMD PR: https://github.com/tenstorrent/tt-umd/pull/1705
Passing CI Runs : All-post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/20969272665
                              Merge gate: https://github.com/tenstorrent/tt-metal/actions/runs/20988390171
## Overview

This PR introduces **Mock Device** - a simulation layer that enables testing and development of tt-metal applications without physical Tenstorrent hardware. Mock Device provides a complete software-only implementation of device operations, allowing developers to validate APIs, test workflows, and develop multi-chip applications entirely in software.

### What is Mock Device?

Mock Device is a hardware emulation layer that simulates Tenstorrent accelerator behavior without requiring physical chips. It provides. Multi-chip configurations can also be emulated by specifying a cluster descriptor YAML. 

### Usage

**Environment Variable approach:**
export TT_METAL_MOCK_CLUSTER_DESC_PATH=tt_metal/third_party/umd/tests/cluster_descriptor_examples/blackhole_P150.yaml

**Programmatic API approach:**
#include <tt-metalium/experimental/mock_device.hpp>

// Configure mock mode BEFORE any device creation
tt::tt_metal::experimental::configure_mock_mode(tt::ARCH::BLACKHOLE, 1);

// Now use MeshDevice::create() as normal - it will automatically use mock mode
auto mesh = tt::tt_metal::distributed::MeshDevice::create(mesh_config, mesh_device_options);

// Alternatively, auto-detect architecture from hardware
// (useful when HW is present but you want mock mode for graph capture)
tt::tt_metal::experimental::configure_mock_mode_from_hw();The programmatic API sets the mock configuration before `MetalContext` initializes, so all subsequent device operations (via `MeshDevice::create()`) will use the mock backend transparently.

### Supported Configurations

| Arch | Chips | Descriptor |
|------|-------|------------|
| WORMHOLE_B0 | 1, 2, 4, 8, 32 | N150, N300, 2x2, T3K, TG |
| BLACKHOLE | 1, 2 | P150, P300 |

### Additional Fixes

Fixed pre-existing Unity Build issues: missing `#pragma once` in `device_command_calculator.hpp` 

### Next Steps
- Integration into L2-nightly-pipeline for regression testing

abhullar-tt reviewed (commented):
I think perhaps we can look into adding MockCommandQueue / MockSystemMemoryManager so we don't need these guards everywhere

abhullar-tt commented on tt_metal/impl/device/device.cpp:
we should have this default true because we generally expect all device to be using translated coordinates

msudumbrekar-TT commented on tt_metal/impl/dispatch/system_memory_manager.cpp:
Comments removed.

msudumbrekar-TT commented on tests/tt_metal/tt_metal/test_mock_device.cpp:
```diff
@@ -0,0 +1,133 @@
+};
+
+// Helper to run existing test binaries as subtests
+class MockDeviceTestRunner : public MockDeviceTestFixture {
```
Removed!

fvranicTT commented on tt_metal/third_party/tt_llk:
```diff
@@ -1 +1 @@
-Subproject commit dfd8dc0bf45e3cbd6c57eb38109bb46449a1609f
+Subproject commit 8a889e55bed288c48a0f4cfa5ed671519bbc33fc
```
Please revert changes in the submodules (umd, llk and tracy).

abhullar-tt commented on tt_metal/fabric/fabric_context.cpp:
```diff
@@ -190,5 +190,135 @@ bool FabricContext::need_deadlock_avoidance_support(eth_chan_directions directio
 
+size_t FabricContext::get_fabric_packet_header_size_bytes() const { return this->packet_header_size_bytes_; }
+
+size_t FabricContext::get_fabric_max_payload_size_bytes() const { return this->max_payload_size_bytes_; }
```
same here

abhullar-tt commented on tt_metal/impl/buffers/dispatch.cpp:
```diff
@@ -712,6 +712,8 @@ void write_to_device_buffer(
     tt::stl::Span<const uint32_t> expected_num_workers_completed,
     CoreType dispatch_core_type,
     tt::stl::Span<const SubDeviceId> sub_device_ids) {
+    ZoneScoped;
```
I think this should be removed

abhullar-tt commented on tt_metal/impl/device/device.cpp:
```diff
@@ -649,6 +656,19 @@ CommandQueue& Device::command_queue(std::optional<uint8_t> cq_id) {
+        if (mock_sysmem_managers.find(this->id_) == mock_sysmem_managers.end()) {
+            mock_sysmem_managers[this->id_] = std::make_unique<SystemMemoryManager>(this->id_, 1);
+        }
+        return *mock_sysmem_managers[this->id_];
```
in this mock case why is this an array vs real device there is just one? and why does this need to be created separately since the SystemMemoryManager automatically stubs for mock?

abhullar-tt reviewed (commented):
Can you update L2-nightly cpp job to include running dispatch test suite with MockDevice env var in this PR. Once that is done could you add a link to passing post commit and the L2 nightly on the PR?

abhullar-tt commented on run_mock_device_tests.sh:
```diff
@@ -0,0 +1,56 @@
+# Run allocator tests (Priority 1 - Original requirement)
+echo ""
+echo "=== Testing Allocators (Fast Dispatch) ==="
+./build_Release/test/tt_metal/unit_tests_api --gtest_filter="*Allocator*"
```
these aren't interesting with mock device since they don't do on device operations. Would remove this

abhullar-tt reviewed (commented):
nice work to get this through!

abhullar-tt commented on tt_metal/impl/context/metal_context.cpp:
```diff
@@ -312,25 +313,35 @@ void MetalContext::teardown() {
     }
 
-    watcher_server_->detach_devices();
+    // Skip watcher detach for mock devices (they may not have been properly attached)
```
could you remove this comment, its not required and seems ai picked up wrong context (we wouldn't attach watcher in mock cases)

abhullar-tt commented on tt_metal/impl/device/device_manager.cpp:
```diff
@@ -397,7 +400,10 @@ void DeviceManager::initialize_active_devices() {
     tt_fabric::FabricConfig fabric_config = tt::tt_metal::MetalContext::instance().get_fabric_config();
     if (tt_fabric::is_tt_fabric_config(fabric_config)) {
-        if (has_flag(
+        // Mock devices don't have real fabric hardware, skip fabric initialization
```
remove comment

abhullar-tt commented on tt_metal/impl/program/program.cpp:
```diff
@@ -1450,15 +1455,29 @@ void detail::ProgramImpl::compile(IDevice* device, bool force_slow_dispatch) {
+                        kernel->register_kernel_elf_paths_with_watcher(*device);
+                    }
+
+                    // Mock devices don't compile kernels - skip binary generation and use empty stubs
```
remove comment

abhullar-tt commented on tt_metal/llrt/tt_cluster.cpp:
```diff
@@ -376,6 +376,7 @@ const std::unordered_map<CoreCoord, int32_t>& Cluster::get_virtual_routing_to_pr
 void Cluster::open_driver(const bool& /*skip_driver_allocs*/) {
     std::unique_ptr<tt::umd::Cluster> device_driver;
     std::string sdesc_path = get_soc_description_file(this->arch_, this->target_type_, rtoptions_);
+    std::unique_ptr<umd::ClusterDescriptor> mock_cluster_desc;  // Declare here to ensure it lives for entire function
```
remove comment here

msudumbrekar-TT committed:
Add programmatic API for mock device mode

- Add configure_mock_mode(arch, num_chips) API for tt-mlir/tt-xla
- Add configure_mock_mode_from_hw() for auto-detection
- Add disable_mock_mode() to turn off mock mode
- Integrate with MetalContext to check for programmatic config
- Add gtest for API verification
- No environment variables needed to enable mock mode

msudumbrekar-TT committed:
Address review comments and fixes

msudumbrekar-TT commented on tt_metal/impl/device/mock_device.cpp:
```diff
@@ -0,0 +1,80 @@
+    return g_registered_mock_config;
+}
+
+std::string get_mock_cluster_desc_path(const MockDeviceConfig& config) {
```
Made the change

msudumbrekar-TT commented:
/codeowners ping

msudumbrekar-TT commented on tt_metal/impl/device/mock_device.cpp:
```diff
@@ -0,0 +1,80 @@
+    configure_mock_mode(arch, num_chips);
+}
+
+void disable_mock_mode() {
```
Addressed.

fvranicTT commented on tt_metal/impl/program/program.cpp:
```diff
@@ -1452,12 +1469,17 @@ void detail::ProgramImpl::compile(IDevice* device, bool force_slow_dispatch) {
+    bool is_mock =
+        tt::tt_metal::MetalContext::instance().get_cluster().get_target_device_type() == tt::TargetDevice::Mock;
+    if (!is_mock) {
+        for (auto& kernels : kernels_) {
```
can these be const?
Also, you are not using id, so you could avoid structured bindings to silence the compiler complaining about unused variable with:
```cpp
for (const auto& kernels : kernels_) {
    for (const auto& pair : kernels) {
        const auto& kernel = pair.second;  // Only extract what you need
        launch_build_step([kernel, device] { kernel->read_binaries(device); }, events);
    }
}
```

jbaumanTT commented on tt_metal/distributed/fd_mesh_command_queue.cpp:
```diff
@@ -129,6 +129,10 @@ FDMeshCommandQueue::FDMeshCommandQueue(
 }
 
 FDMeshCommandQueue::~FDMeshCommandQueue() {
+    // Mock devices don't have actual completion queues, skip validation
```
I'd recommend adding similar changes to sd_mesh_command_queue.cpp as well.

msudumbrekar-TT replied:
Made the change!

jbaumanTT replied:
Sorry, could you also adds skips to SDMeshCommandQueue::enqueue_mesh_workload, enqueue_write_buffer, enqueue_read_buffer, etc?

msudumbrekar-TT replied:
Done!

msudumbrekar-TT commented on tt_metal/impl/dispatch/system_memory_manager.cpp:
```diff
@@ -268,29 +314,47 @@ bool SystemMemoryManager::get_bypass_mode() const { return this->bypass_enable;
 
 uint32_t SystemMemoryManager::get_completion_queue_limit(const uint8_t cq_id) const {
+    if (is_mock_device()) {
+        return 0x10000;
```
Done!

msudumbrekar-TT commented on tt_metal/llrt/rtoptions.hpp:
```diff
@@ -641,6 +641,17 @@ class RunTimeOptions {
+    void set_mock_cluster_desc(const std::string& filename) {
+        if (!filename.empty()) {
+            mock_cluster_desc_path =
+                get_root_dir() + "/tt_metal/third_party/umd/tests/cluster_descriptor_examples/" + filename;
```
Just added a comment! Should be fine with this trade-off for now! Thanks!

fvranicTT commented on tests/tt_metal/tt_metal/dispatch/dispatch_program/test_dispatch.cpp:
```diff
@@ -352,32 +352,36 @@ TEST_F(MeshDispatchFixture, TensixActiveEthTestCBsAcrossDifferentCoreTypes) {
+                EXPECT_TRUE(pass_intermediate);
+
+                uint32_t out_index = out_cb * sizeof(uint32_t);
+                bool addr_match_out = cb_config_vector.at(out_index) == cb_addr;
```
Here too.

fvranicTT commented on tt_metal/impl/context/metal_context.cpp:
```diff
@@ -428,6 +443,13 @@ void MetalContext::teardown_base_objects() {
 
 MetalContext::MetalContext() {
+    // Check if mock mode was configured via API (before env vars take effect)
+    auto mock_cluster_desc = experimental::get_mock_cluster_desc();
```
You could use if with initializer:
```cpp
if (auto mock_cluster_desc = experimental::get_mock_cluster_desc()) {
    rtoptions_.set_mock_cluster_desc(*mock_cluster_desc);
    log_info(tt::LogMetal,
             "Using programmatically configured mock mode: {}",
             *mock_cluster_desc);
}
```

fvranicTT commented on tt_metal/impl/context/metal_context.cpp:
```diff
@@ -1022,7 +1070,9 @@ void MetalContext::generate_device_bank_to_noc_tables(ChipId device_id) {
     dram_bank_to_noc_xy_[device_id].clear();
     dram_bank_to_noc_xy_[device_id].reserve(hal_->get_num_nocs() * num_dram_banks);
-    bool noc_translation_enabled = cluster_->get_cluster_desc()->get_noc_translation_table_en().at(device_id);
+    bool noc_translation_enabled = (cluster_->get_target_device_type() == tt::TargetDevice::Mock)
```
```cpp
bool noc_translation_enabled =
    cluster_->get_target_device_type() != tt::TargetDevice::Mock &&
    cluster_->get_cluster_desc()
        ->get_noc_translation_table_en()
        .at(device_id);
```

msudumbrekar-TT commented on tests/tt_metal/tt_metal/dispatch/dispatch_program/test_dispatch.cpp:
```diff
@@ -352,32 +352,36 @@ TEST_F(MeshDispatchFixture, TensixActiveEthTestCBsAcrossDifferentCoreTypes) {
+                uint32_t cb_addr = mesh_device->allocator()->get_base_allocator_addr(HalMemType::L1);
+                uint32_t intermediate_index = intermediate_cb * sizeof(uint32_t);
+
+                bool addr_match_intermediate = cb_config_vector.at(intermediate_index) == cb_addr;
```
Made the change!

msudumbrekar-TT commented on tt_metal/impl/device/mock_device.cpp:
```diff
@@ -0,0 +1,73 @@
+
+    const auto& config = *g_registered_mock_config;
+
+    if (config.arch == tt::ARCH::WORMHOLE_B0) {
```
Good point! Made the change!

msudumbrekar-TT commented:
/ping @jbaumanTT

akerteszTT commented on tt_metal/api/tt-metalium/experimental/mock_device.hpp:
```diff
@@ -0,0 +1,34 @@
+#include <string>
+
+namespace tt::tt_metal::experimental {
+
```
Maybe consider throwing in a comment at the top of this API file to explain what this is for. Is it intended to be for just internal use?

I'm easy either way -- that's what `experimental` is here for.

akerteszTT reviewed (approved):
Thanks for putting this in `experimental`!

Riddy21 commented on tt_metal/impl/context/metal_context.cpp:
```diff
@@ -781,6 +815,13 @@ tt_fabric::FabricUDMMode MetalContext::get_fabric_udm_mode() const { return fabr
+        control_plane_ = std::make_unique<tt::tt_fabric::ControlPlane>();
+        return;
+    }
+
```
When using mock device, there is still control plane verification which needs to happen. Specifically for the Mesh Graph Descriptor. So we must use `control_plane_ = std::make_unique<tt::tt_fabric::ControlPlane>(mesh_graph_desc_path.string());`.

Otherwise all tests using Mock device and mesh graph descriptor will fail in CPU tests: `tt-metal/.github/workflows/fabric-cpu-only-tests-impl.yaml`

You can choose not to use data from control plane, but control plane must still be initialized properly

Riddy21 reviewed (changes requested):
Excited for this change! Think this will be a good change to get more CPU only testing coverage.

However, couple things that need to be fixed in control plane or else this will fail merge gate. Highlighted them below.

You can test locally with running [these tests](https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/fabric-cpu-only-tests-impl.yaml#L58-L129)

msudumbrekar-TT commented on tt_metal/impl/context/metal_context.cpp:
```diff
@@ -791,6 +832,13 @@ void MetalContext::construct_control_plane(const std::filesystem::path& mesh_gra
+        control_plane_ = std::make_unique<tt::tt_fabric::ControlPlane>();
+        return;
+    }
+
```
Makes sense! Removed!

msudumbrekar-TT commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -1098,6 +1098,13 @@ FabricNodeId ControlPlane::get_fabric_node_id_from_physical_chip_id(ChipId physi
+    if (tt::tt_metal::MetalContext::instance().get_cluster().get_target_device_type() == tt::TargetDevice::Mock) {
+        return FabricNodeId(MeshId{0}, physical_chip_id);
+    }
+
```
Tested removing this - it breaks large mock clusters. The smaller configs work without it, but larger ones need this fallback.

msudumbrekar-TT commented on tt_metal/llrt/rtoptions.hpp:
```diff
@@ -637,6 +637,20 @@ class RunTimeOptions {
+        if (simulator_path.empty()) {
+            runtime_target_device_ = tt::TargetDevice::Mock;
+        }
+    }
```
Good point -this is designed to be called before MetalContext::initialize(). The cluster reads the path during initialization via get_mock_cluster_desc(rtoptions_). Added a comment clarifying the intended call order.

blozano-tt committed:
Reorder tile.cpp inclusion in CMakeLists.txt

msudumbrekar-TT committed:
Add mock device support

msudumbrekar-TT committed:
Move mock device test script to tests/scripts/ and remove TT_METAL_HOME dependency

msudumbrekar-TT commented:
/codeowners ping @tenstorrent/metalium-developers-infra
